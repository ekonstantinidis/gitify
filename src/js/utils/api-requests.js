var request = require('superagent');
var AuthStore = require('../stores/auth');

var apiRequests = {
  get: function (url) {
    return request
      .get(url)
      .set('Accept', 'application/json');
  },

  post: function (url, params) {
    return request
      .post(url)
      .send(params)
      .set('Accept', 'application/json')
      .set('User-Agent', 'Gitify');
  },

  getAuth: function (url) {
    return request
      .get(url)
      .set('Accept', 'application/vnd.github.v3+json')
      .set('Authorization', 'token ' + AuthStore.authStatus())
      .set('Cache-Control', 'no-cache');
  },

  putAuth: function (url, params) {
    return request
      .put(url)
      .send(params)
      .set('Accept', 'application/vnd.github.v3+json')
      .set('Authorization', 'token ' + AuthStore.authStatus())
      .set('Cache-Control', 'no-cache');
  },

  patchAuth: function (url, params) {
    return request
      .patch(url)
      .send(params)
      .set('Accept', 'application/vnd.github.v3+json')
      .set('Authorization', 'token ' + AuthStore.authStatus())
      .set('Cache-Control', 'no-cache');
  }
};

module.exports = apiRequests;
