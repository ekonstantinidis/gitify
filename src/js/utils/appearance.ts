import { Appearance } from '../../types';

export const setAppearance = (mode: Appearance) => {
  const htmlClassList = document.querySelector('html').classList;
  const setLightMode = () => htmlClassList.remove('dark');
  const setDarkMode = () => htmlClassList.add('dark');

  switch (mode) {
    case Appearance.LIGHT:
      setLightMode();
      break;

    case Appearance.DARK:
      setDarkMode();
      break;

    default:
      if (
        window.matchMedia &&
        window.matchMedia('(prefers-color-scheme: dark)').matches
      ) {
        setDarkMode();
      } else {
        setLightMode();
      }
  }
};
