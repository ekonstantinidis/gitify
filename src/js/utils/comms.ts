const { ipcRenderer, remote, shell } = require('electron');

import { SettingsState } from '../../types/reducers';

export function openExternalLink(url) {
  shell.openExternal(url);
}

export function setAutoLaunch(value) {
  remote.app.setLoginItemSettings({
    openAtLogin: value,
    openAsHidden: value,
  });
}

export function updateTrayIcon(notificationsLength = 0) {
  if (notificationsLength > 0) {
    ipcRenderer.send('update-icon', 'TrayActive');
  } else {
    ipcRenderer.send('update-icon');
  }
}

export function reOpenWindow() {
  ipcRenderer.send('reopen-window');
}

export function restoreSetting(setting, value) {
  ipcRenderer.send(setting, value);
}
