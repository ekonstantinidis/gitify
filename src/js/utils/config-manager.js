var config = require('../../../config');

module.exports = {
  config: config,
  hostUrl: function getHostUrl() {
    return config.github.proto + '://' + config.github.host;
  },
  apiUrl: function getApiUrl() {
    return config.github.host.indexOf('github.com') === -1 ?
      this.hostUrl() + '/api/' + config.github.apiVersion :
      config.github.proto + '://api.' + config.github.host;
  },
  apiHost: function getApiHostOnly() {
    return this.apiUrl().replace('https://', '').replace('http://', '');
  }
};
