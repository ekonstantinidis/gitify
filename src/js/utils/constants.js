const constants = {
  SCOPE: ['user:email', 'notifications'],

  REPO_SLUG: 'manosim/gitify',

  // Storage
  STORAGE_KEY: 'gitify-storage',

  // Awesome all read messages
  ALLREAD_MESSAGES: [
    'Wow! You did it.',
    'That\'s amazing!',
    'Yes! All read.',
    'All gone! Nice work!',
    'Yay! Good news.',
  ],

  ALLREAD_EMOJIS: [
    ':wink:',
    ':tada:',
    ':tiger:',
    ':see_no_evil:',
    ':balloon:',
    ':confetti_ball:',
    ':clap:',
    ':circus_tent:',
    ':spaghetti:',
    ':ok_hand:'
  ],

  ERROR_EMOJIS: [
    ':pensive:',
    ':disappointed:',
    ':triumph:',
    ':scream:',
    ':cry:'
  ]
};

export default constants;
