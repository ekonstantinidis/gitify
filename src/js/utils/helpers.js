var electron = require('electron');
var remote = electron.remote;
var BrowserWindow = remote.BrowserWindow;

import Constants from './constants';

export default {
  generateGitHubUrl(isEnterprise, url) {
    let newUrl = isEnterprise
      ? url.replace('github.intuit.com/api/v3/repos', 'github.intuit.com')
      : url.replace('api.github.com/repos', 'www.github.com');

    if (newUrl.indexOf('/pulls/') !== -1) {
      newUrl = newUrl.replace('/pulls/', '/pull/');
    }

    if (newUrl.indexOf('/releases/') !== -1) {
      newUrl = newUrl.replace('/repos', '');
      newUrl = newUrl.substr(0, newUrl.lastIndexOf('/'));
    }

    return newUrl;
  },

  authGithub(settings, loginUser) {
    //Build the OAuth consent page URL
    var authWindow = new BrowserWindow({
      width: 800,
      height: 600,
      show: true,
      webPreferences: {
        nodeIntegration: false
      }
    });
    var githubUrl = `https://${settings.get('baseUrl')}/login/oauth/authorize?`;
    var authUrl = githubUrl + 'client_id=' + settings.get('clientId') + '&scope=' + Constants.SCOPE;
    authWindow.loadURL(authUrl);

    function handleCallback (url) {
      var raw_code = /code=([^&]*)/.exec(url) || null;
      var code = (raw_code && raw_code.length > 1) ? raw_code[1] : null;
      var error = /\?error=(.+)$/.exec(url);

      if (code || error) {
        // Close the browser if code found or error
        authWindow.destroy();
      }

      // If there is a code, proceed to get token from github
      if (code) {
        loginUser(code);
      } else if (error) {
        alert('Oops! Something went wrong and we couldn\'t ' +
          'log you in using Github. Please try again.');
      }
    }

    // If "Done" button is pressed, hide "Loading"
    authWindow.on('close', function () {
      authWindow.destroy();
    });

    authWindow.webContents.on('will-navigate', function (event, url) {
      handleCallback(url);
    });

    authWindow.webContents.on('did-get-redirect-request', function (event, oldUrl, newUrl) {
      handleCallback(newUrl);
    });
  }
};
