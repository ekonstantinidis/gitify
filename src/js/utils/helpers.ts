import { parse } from 'url';
const { remote } = require('electron');
const BrowserWindow = remote.BrowserWindow;
const dialog = remote.dialog;

import Constants from './constants';
import { loginUser } from '../actions';
import { AuthState } from '../../types/reducers';

export function getEnterpriseAccountToken(hostname, accounts): string {
  return accounts.find((obj) => obj.hostname === hostname).token;
}

export function generateGitHubAPIUrl(hostname) {
  const isEnterprise = hostname !== Constants.DEFAULT_AUTH_OPTIONS.hostname;
  return isEnterprise
    ? `https://${hostname}/api/v3/`
    : `https://api.${hostname}/`;
}

export function generateGitHubWebUrl(url: string) {
  const { hostname } = parse(url);
  const isEnterprise =
    hostname !== `api.${Constants.DEFAULT_AUTH_OPTIONS.hostname}`;

  let newUrl: string = isEnterprise
    ? url.replace(`${hostname}/api/v3/repos`, hostname)
    : url.replace('api.github.com/repos', 'github.com');

  if (newUrl.indexOf('/pulls/') !== -1) {
    newUrl = newUrl.replace('/pulls/', '/pull/');
  }

  if (newUrl.indexOf('/releases/') !== -1) {
    newUrl = newUrl.replace('/repos', '');
    newUrl = newUrl.substr(0, newUrl.lastIndexOf('/'));
  }

  return newUrl;
}

export function authGithub(
  authOptions = Constants.DEFAULT_AUTH_OPTIONS,
  dispatch
) {
  // Build the OAuth consent page URL
  const authWindow = new BrowserWindow({
    width: 800,
    height: 600,
    show: true,
  });

  const githubUrl = `https://${authOptions.hostname}/login/oauth/authorize`;
  const authUrl = `${githubUrl}?client_id=${authOptions.clientId}&scope=${Constants.AUTH_SCOPE}`;

  const session = authWindow.webContents.session;
  session.clearStorageData();

  authWindow.loadURL(authUrl);

  function handleCallback(url) {
    const raw_code = /code=([^&]*)/.exec(url) || null;
    const code = raw_code && raw_code.length > 1 ? raw_code[1] : null;
    const error = /\?error=(.+)$/.exec(url);

    if (code || error) {
      // Close the browser if code found or error
      authWindow.destroy();
    }

    // If there is a code, proceed to get token from github
    if (code) {
      dispatch(loginUser(authOptions, code));
    } else if (error) {
      alert(
        "Oops! Something went wrong and we couldn't " +
          'log you in using Github. Please try again.'
      );
    }
  }

  // If "Done" button is pressed, hide "Loading"
  authWindow.on('close', () => {
    authWindow.destroy();
  });

  authWindow.webContents.on(
    'did-fail-load',
    (event, errorCode, errorDescription, validatedURL) => {
      if (validatedURL.includes(authOptions.hostname)) {
        authWindow.destroy();
        dialog.showErrorBox(
          'Invalid Hostname',
          `Could not load https://${authOptions.hostname}/.`
        );
      }
    }
  );

  authWindow.webContents.on('will-redirect', (event, url) => {
    event.preventDefault();
    handleCallback(url);
  });
}

export function isUserEitherLoggedIn(auth: AuthState) {
  return auth.token !== null || auth.enterpriseAccounts.length > 0;
}
