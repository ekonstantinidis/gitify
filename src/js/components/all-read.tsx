import * as React from 'react';
import Typist from 'react-typist';
import { emojify } from 'react-emojione';

import { Constants } from '../../utils/Constants';

export const AllRead = () => {
  const message = React.useMemo(
    () =>
      Constants.ALLREAD_MESSAGES[
        Math.floor(Math.random() * Constants.ALLREAD_MESSAGES.length)
      ],
    []
  );

  const emoji = React.useMemo(
    () =>
      Constants.ALLREAD_EMOJIS[
        Math.floor(Math.random() * Constants.ALLREAD_EMOJIS.length)
      ],
    []
  );

  return (
    <div className="flex flex-1 flex-col justify-center items-center p-4 bg-white dark:bg-gray-dark text-black dark:text-white">
      <h1 className="text-5xl mb-5">{emojify(emoji, { output: 'unicode' })}</h1>

      <h2 className="font-semibold text-xl mb-2 text-semibold">
        <Typist>{message}</Typist>
      </h2>

      <div>No new notifications.</div>
    </div>
  );
};
