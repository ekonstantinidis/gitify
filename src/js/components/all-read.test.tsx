import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { Constants } from '../../utils/Constants';

import { AllRead } from './all-read';

jest.mock('react-typist');

describe('components/all-read.tsx', function () {
  it('should render itself & its children', function () {
    spyOn(Constants, 'ALLREAD_EMOJIS');
    spyOn(Constants, 'ALLREAD_MESSAGES');

    const tree = TestRenderer.create(<AllRead />);

    expect(tree).toMatchSnapshot();
  });
});
