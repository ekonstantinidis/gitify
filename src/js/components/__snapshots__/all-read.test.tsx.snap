// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/all-read.tsx should render itself & its children 1`] = `
<div
  className="flex flex-1 flex-col justify-center items-center p-4"
>
  <h1
    className="text-5xl mb-5"
  >
    ðŸ˜‰
  </h1>
  <h2
    className="font-semibold text-xl mb-2 text-semibold"
  />
  <div>
    No new notifications.
  </div>
</div>
`;
